Threads to be sent to municipality

PetrDlouhy opened the pull request:
This is unfinished functionality prototype meant to be discussed and finished before pulling.

This PR adds new button "Send to the municipality" which adds new message thread with parameter that it should be send to the municipality. We are working in cooperation with Prague city council on this function which will send automatically Issue data to the official portal [http://zmente.to](http://zmente.to).
The new thread form is prefilled with the issue title and description, the user can modify this or leave it.
This function is available only for members of the groups to prevent inexperienced users from sending low quality issues.

There some things, that has to be solved before pulling:
- The administrator should be able to add new external services user can choose from.
- If there is no such service, user should not see the button.

Please comment, what do you think about this function.

PetrDlouhy committed:
add external_service parameter to thread API

PetrDlouhy committed:
add send to municipality button

nikolai-b commented:
Wow, that is an exciting possibility.  I think it is really for @mvl22 to deal with as this is not a coding issue so I can't comment.

mvl22 commented:
There seem to be hard-coded references to the specific service (zmente.to) in the code - I think this stuff should be generalised, perhaps in the form of a driver file which implements particular services and then the settings contain a drop down of the available drivers (albeit only one at present).

On a policy question, if any member can submit an issue externally, what is required to ensure that the submission represents the considered view of the members of the discussion thread as a whole? For instance, couldn't a new user come along, and submit something wildly different to the consensus view?

PetrDlouhy commented:
@mvl22 For now, this only sets `submit_external` parameter of message thread to the service identifier value (`zmenteto`). Then the marked threads can be filtered in the API. I solved the sending from Cyclescape API to Změňte.to API by external script. Although it might be desired to include external service senders to the code in the future, I went the simplest way for me now.

I would like to add a possibility to configure possible services in the site config. My idea is, that user only fills service name and service identifier (and maybe other identifiers in the future) to the table. If the table is empty, no `Send to municipality` button is shown.
But I am not quite sure how to do this. The only way I can imagine is changing the `submit_external` parameter to foreign key to the service table.

Změňte.to is a service opened to the wide public, so opening this possibility to the group member is more restrictive than the original service. So for my use-case, this is OK and I would like to keep this simple. If anyone wants to add some service that requires more restrictive approach, he can add it by some settings of the services.

PetrDlouhy commented:
I updated this PR and I consider it to be feature complete now. I changed the thread parameter to `external_service` foreign key and added external services configuration to admin.

So now there is no reference to any particular service and everything is done by configuration in admin. If no service is added in admin, no "Send to municipality" button is displayed, so nothing changes.

Please review.

PetrDlouhy commented:
I rebased this to `cs/staging`

nikolai-b commented:
Thanks @PetrDlouhy I don't have time to review this one properly yet and I'd like to do it after the next release, do you mind it waiting here for a bit?

Thanks

PetrDlouhy commented:
I would like, if it got there, but I cannot force you :-)

PetrDlouhy commented:
@nikolai-b When will you have time to review this? If you write me in advance, I could rebase this before your review.

mvl22 commented:
> I could rebase this before your review.

Yes, if you could update to the latest code, that will speed up review - I see there is a conflict noted above ^^^.

nikolai-b commented on db/migrate/20171212101253_add_submit_external_to_message_thread.rb:
```diff
@@ -0,0 +1,9 @@
+class AddSubmitExternalToMessageThread < ActiveRecord::Migration
+  def change
+    add_column :message_threads, :external_service_id, :integer
```
If we use 
`add_reference :message_threads, :external_service, foreign_key: true`
it adds and index and preserves referential integrity (i.e. we can't have a thread referencing a non-existent external service.

nikolai-b commented on db/migrate/20171212101253_add_submit_external_to_message_thread.rb:
```diff
@@ -0,0 +1,9 @@
+    create_table :external_services do |t|
+      t.string :name, null: false
+      t.string :short_name, null: false
+    end
```
We want a uniqueness constraint on `:short_name`
`add_index :external_services, :short_name, unique: true`

nikolai-b commented on db/migrate/20171212101253_add_submit_external_to_message_thread.rb:
```diff
@@ -0,0 +1,9 @@
+      t.string :short_name, null: false
+    end
+  end
+end
```
In pulls it helps if you also run `rake db:migrate` and commit the schema changes.

nikolai-b commented on app/models/external_service.rb:
```diff
@@ -0,0 +1,27 @@
+  has_many :threads, class_name: 'MessageThread', inverse_of: :external_service
+
+  validates :name, presence: true, uniqueness: true
+  validates :short_name, presence: true, uniqueness: true, subdomain: true
```
What is this?  As in what is the difference between the name and short name?  Why not just use the name?

nikolai-b reviewed (commented):
I'm unsure about the UI change. 

![screenshot from 2018-09-17 19-21-16](https://user-images.githubusercontent.com/6289830/45642139-54d3a600-baaf-11e8-8251-c8c9b04e6c1b.png)

Which creates a normal new thread and marks it as external which presumably your script then is able to pull in from the API.  Why not just add it as a drop down (but allow it to be blank) whenever a new thread is created?

Other than that the code is fine, I've made a few minor comments.  Thank you @PetrDlouhy !

mvl22 commented on app/views/home/user_profile_template.html.erb:
```diff
@@ -79,7 +83,9 @@
-            <div class="permissions">shared</div>
+            <div class="thread-parameters">
+               <div class="permissions">shared</div>
+            </div>
```
> was there a reason the original designs were kept as part of this repository

Because the design has never actually been fully implemented yet! That said, I am trying to get round to commissioning a fresh new design that mops up piles of usability issues.

nikolai-b commented on app/controllers/issue/message_threads_controller.rb:
```diff
@@ -17,6 +17,24 @@ def new
+    end
+    @message = @thread.messages.build
+    @message.body = issue.description
+    @thread.title = issue.title
```
I'm not sure about setting the thread title equal to the issue title.
I think it should be:
Issue: New housing will be built
Thread: Increased congestion on the road
Thread: Cycle Parking on the New housing application is poor

etc.

See https://github.com/cyclestreets/cyclescape/issues/43 for context.

mvl22 commented:
> which adds new message thread with parameter that it should be send to the municipality. We are working in cooperation with Prague city council on this function which will send automatically Issue data to the official portal http://zmente.to.

To be clear - is this purely copying the initial problem report to the municipality?

nikolai-b commented:
@mvl22 is https://github.com/cyclestreets/cyclescape/pull/741#issuecomment-422127191 addressed to me or @PetrDlouhy ?  I think the Send To Municipality button will make a new thread and mark it as such so the thread is sent to the external service.

mvl22 commented:
> @mvl22 is #741 (comment) addressed to me or @PetrDlouhy?

To Petr.

> I think the Send To Municipality button will make a new thread and mark it as such so the thread is sent to the external service.

Ah, this is not what I was expecting. Peter - why does a report also create a new thread? Can these not be separated into "Create a new thread" (which already exists) and "Send to municipality" (as to be added)? I think it would be clearest to keep these two things completely separate.
